keps/sig-apps/0012-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,162 @@
+            * [Update](#update)
+            * [Rollback](#rollback)
+         * [Alternatives](#alternatives)
+            
+## Summary
+We propose a new update strategy for deployment controller to do rolling update without tearing down the pods, i.e. in-place update.
+
+
+## Motivation
+Today, Deployment controller, while doing rolling update, will teardown all the pods and then recreate all the pods.
+This approach works in most cases but has several limitations:
+* The re-created pod may be allocated onto different nodes, but certain pods depend on local state on the node. 
+
+    For example: some pods have persistent state in local storage. 
+    Especially when we do high pressure with very large scale testing, we may go through several rounds of upgrades. If the pods randomly run on any nodes, 
+    the network environment is not consistent, it becomes hard to guarantee a stable result.
+* The new pod may not be allocated due to resources taken by other pods. Preemption can mitigate this situation, but this 
+introduces an extra scheduling loop that can be avoided.
+* The IP address may be changed due to the pod is re-created, but we want fixed IP across container upgrades. 
+
+    It is true that pods in Kubernetes don't have sticky IPs by default. In the scenario where the node fails, 
+    we implemented a CNI plugin to ensure that the new pod allocated still get the same ip address.
+
+
+## Goals
+Enable deployment controller to update the containers inside the pods (such as container image, cmd etc.) without tearing down the pods
+
+## Proposal
+
+### API Change
+Augment deployment controller to have a new update strategy type: InPlaceUpdate.
+
+```
+
+type DeploymentStrategy struct {
+	// Type of deployment. Can be "Recreate" or "RollingUpdate" or "InPlaceUpdate". Default is RollingUpdate.
+	// +optional
+	Type DeploymentStrategyType
+
+	// Rolling update config params. Present only if DeploymentStrategyType =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be.
+	// +optional
+	RollingUpdate *RollingUpdateDeployment
+
++	// InPlace update config params. Present only if DeploymentStrategyType =
++	// InPlaceUpdate.
++	// +optional
++	InPlaceUpdate *InPlaceUpdateDeployment
+}
+
+const(
+	// Kill all existing pods before creating new ones.
+	RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
+
+        // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+        RollingUpdateDeploymentStrategyType DeploymentStrategyType = "RollingUpdate"
+    
+        // Update containers inside the pods without tearing down the pods
++	InPlaceUpdateDeploymentStrategyType DeploymentStrategyType = "InPlaceUpdate"
+)
+
+// Spec to control the desired behavior of in-place update
++ type InPlaceUpdateDeployment struct {
++	 // The number of pods to be updated in each batch.
```

Comment:
Make explicit in the doc that this is gated by readiness.

Reply:
updated.